File: 天梯赛/L2-023. 图着色问题.cpp
#include<iostream>
#include<set>
using namespace std;
const int maxn=510;
bool e[maxn][maxn]={false};
int color[maxn];
int main(){
	int m,n,k,a,b,q;
	cin>>n>>m>>k;
	for(int i=0;i<m;++i){
		cin>>a>>b;
		e[a][b]=e[b][a]=true;
	}
	cin>>q;
	while(q--){
		bool flag=true;
		set<int> s;
		for(int i=1;i<=n;++i){
			cin>>color[i];
			s.insert(color[i]);
		}
		if(k!=s.size())flag=false;
		for(int i=1;i<=n;++i){
			if(!flag)break;
			for(int j=i+1;j<=n;++j){
				if(e[i][j]&&color[i]==color[j]){
					flag=false;
					break;
				}
			}
		}
		if(flag)cout<<"Yes"<<endl;
		else cout<<"No"<<endl;
	}
	return 0;
}

File: 天梯赛/java/L2-023. 图着色问题.java
import java.util.Arrays;
import java.util.HashSet;
import java.util.Scanner;
import java.util.Set;

public class Main {
	public static final int maxn=501;
	public static boolean e[][]=new boolean [maxn][maxn];
	public static int color[]=new int [maxn];
	public static void main(String[] args) {
		Scanner in=new Scanner(System.in);
		int n=in.nextInt(),m=in.nextInt(),k=in.nextInt();
		for(int i=0;i<m;++i) {
			int a=in.nextInt(),b=in.nextInt();
			e[a][b]=e[b][a]=true;
		}
		int q=in.nextInt();
		for(int cas=0;cas<q;++cas) {
			boolean flag=true;
			Set<Integer> s=new HashSet<>();
			for(int i=1;i<=n;++i) {
				color[i]=in.nextInt();
				s.add(color[i]);
			}
			if(k!=s.size())flag=false;
			for(int i=1;i<=n;++i) {
				if(!flag)break;
				for(int j=i+1;j<=n;++j)
					if(e[i][j]&&color[j]==color[i]) {
						flag=false;
						break;
					}
			}
			if(flag)System.out.println("Yes");
			else System.out.println("No");
		}
	}


}

